payjoin/src/receive/mod.rs

```diff
@@ -428,18 +428,43 @@ impl WantsInputs {
         }
 
         if self.payjoin_psbt.outputs.len() > 2 {
-            // This UIH avoidance function supports only
-            // many-input, n-output transactions such that n <= 2 for now
-            return Err(SelectionError::from(InternalSelectionError::TooManyOutputs));
-        }
-
-        if self.payjoin_psbt.outputs.len() == 2 {
+            // This doesn't attempt to preserve privacy...
+            self.do_coin_selection(candidate_inputs)
+        } else if self.payjoin_psbt.outputs.len() == 2 {
             self.avoid_uih(candidate_inputs)
         } else {
             self.select_first_candidate(candidate_inputs)
         }
     }
 
+    fn do_coin_selection(
+        &self,
+        candidate_inputs: HashMap<Amount, OutPoint>,
+    ) -> Result<Vec<OutPoint>, SelectionError> {
+        // Calculate the amount that the receiver must contribute
+        let output_amount =
+            self.payjoin_psbt.unsigned_tx.output.iter().fold(0, |acc, output| acc + output.value);
+        let original_output_amount =
+            self.original_psbt.unsigned_tx.output.iter().fold(0, |acc, output| acc + output.value);
+        let min_input_amount = min(0, output_amount - original_output_amount);
+
+        // Select inputs that can pay for that amount
+        // TODO: use a better coin selection algorithm
+        let mut selected_coins = vec![];
+        let mut input_sats = 0;
+        for candidate in candidate_inputs {
+            let candidate_sats = candidate.0.to_sat();
+            selected_coins.push(candidate.1);
+            input_sats += candidate_sats;
+
+            if input_sats >= min_input_amount {
+                return Ok(selected_coins);
+            }
+        }
+
+        Err(SelectionError::from(InternalSelectionError::CannotAfford))
+    }
+
```

Comment:
I was thinking of renaming `try_preserving_privacy` based on the discussion in https://github.com/payjoin/rust-payjoin/issues/70, where multiple-input selection was brought up.

But yeah, I agree that it's unnecessary for this PR and enabling cut-through generally. We can delegate input selection to consumers for the non-privacy-preserving cases for now. I'll revert this change.

Reply:
Relatedly, the `change_amount` approach on `WantsInputs` is flawed, because the caller would have to set it correctly themselves and getting it wrong/omitting it would lead to loss of funds. I think it will have to be calculating in `contribute_witness_inputs` by computing (total receiver outputs amount) - (total provided inputs value).

One way to achieve this would be to replace `owned_vouts` with two fields: `receiver_vouts` (all receiver outputs) and `change_vout` (single output to add excess input amount). In the regular payjoin case, `receiver_vouts` == `[change_vout]`.

Reply:
That sounds worth the shot

Reply:
Actually I realized we don't need to keep track of all receiver outputs to calculate the change amount, we can just calculate the amount diff between the proposal PSBT outputs to the original PSBT outputs. So we can keep using just `owned_vouts` as long as we ensure those actually correspond to receiver owned amounts and not forwarded payments.